File: src/main/kotlin/homework/hw07/javaimp/LinkedOrderedList.java
package homework.hw07.javaimp;

import homework.hw07.AbstractOrderedList;

import java.util.Iterator;

/**
 * Linked list implementation of AbstractOrderedList
 *
 * @author Kirill Smirenko, group 271
 */
public class LinkedOrderedList<T extends Comparable<T>> extends AbstractOrderedList<T> {
    private Node head;
    private int size;

    public LinkedOrderedList() {
        head = null;
        size = 0;
    }

    @Override
    public void add(T value) {
        if (head == null) {
            head = new Node(value);
            size++;
            return;
        }
        if (value.compareTo(head.value) < 0) {
            head.next = new Node(head.value, head.next);
            head.value = value;
            size++;
            return;
        }
        Node curNode = head;
        while ((curNode.next != null) && (curNode.next.value.compareTo(value) <= 0)) {
            curNode = curNode.next;
        }
        curNode.next = new Node(value, curNode.next);
        size++;
    }

    @Override
    public T get(int index) {
        if ((index < 0) || (index >= size)) {
            throw new IndexOutOfBoundsException();
        }
        int i = 0;
        Node curNode = head;
        while (i < index) {
            curNode = curNode.next;
            i++;
        }
        return curNode.value;
    }

    @Override
    public void removeAt(int index) {
        if ((index <= 0) || (index >= size)) {
            return;
        }
        int i = 0;
        Node curNode = head;
        while (i < index - 1) {
            curNode = curNode.next;
            i++;
        }
        curNode.next = curNode.next.next;
        size--;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public int hashCode() {
        int h = 0;
        int i = 0;
        Node curNode = head;
        while (i < size) {
            h = 23 * h + curNode.value.hashCode();
            curNode = curNode.next;
            i++;
        }
        return h;
    }

    public Iterator<T> iterator() {
        return new LinkedOrderedListIterator(head);
    }

    private class Node {
        private T value;
        private Node next;

        public Node(T value, Node next) {
            this.value = value;
            this.next = next;
        }

        public Node(T value) {
            this.value = value;
            this.next = null;
        }
    }

    private class LinkedOrderedListIterator implements Iterator<T> {
        private Node node;

        public LinkedOrderedListIterator(Node node) {
            this.node = node;
        }

        public boolean hasNext() {
            return node != null;
        }

        public T next() {
            T val = node.value;
            node = node.next;
            return val;
        }

        public void remove() {
        }
    }
}

File: src/main/kotlin/homework/hw07/kotlinimp/LinkedOrderedList.kt
package homework.hw07.kotlinimp

import homework.hw07.AbstractOrderedList

/**
 * Linked list implementation of AbstractOrderedList

 * @author Kirill Smirenko, group 271
 */
class LinkedOrderedList<T : Comparable<T>> : AbstractOrderedList<T>() {
    private var head : Node<T>? = null
    private var size : Int = 0

    override fun add(value : T) {
        val headSnap = head // to avoid using "!!"
        if (headSnap == null) {
            head = Node(value, null)
            size++
            return
        }
        if (value.compareTo(headSnap.value) < 0) {
            // TODO: CHECK
            headSnap.next = Node(headSnap.value, headSnap.next)
            headSnap.value = value
            //head = headSnap
            size++
            return
        }
        var curNode = headSnap
        while ((curNode?.next != null) && (curNode?.next!!.value.compareTo(value) <= 0)) {
            curNode = curNode?.next
        }
        curNode?.next = Node(value, curNode?.next)
        size++
    }

    override fun get(index : Int) : T {
        if ((index < 0) || (index >= size)) {
            throw IndexOutOfBoundsException()
        }
        var i = 0
        var curNode = head
        while (i < index) {
            curNode = curNode?.next
            i++
        }
        return curNode!!.value
    }

    override fun removeAt(index : Int) {
        if ((index <= 0) || (index >= size)) {
            return
        }
        var i = 0
        var curNode = head
        while (i < index - 1) {
            curNode = curNode?.next
            i++
        }
        curNode?.next = curNode?.next?.next
        size--
    }

    override fun size() : Int {
        return size
    }

    override fun hashCode() : Int {
        var h = 0
        var curNode = head
        while (curNode != null) {
            h = 23 * h + curNode.value.hashCode()
            curNode = curNode.next
        }
        return h
    }

    override fun iterator() : MutableIterator<T> = LinkedOrderedListIterator<T>(head)

    private class LinkedOrderedListIterator<T : Comparable<T>>(private var node : Node<T>?) : MutableIterator<T> {
        override fun hasNext() : Boolean = node != null

        override fun next() : T {
            val nodeSnap = node
            if (nodeSnap != null) {
                val value = nodeSnap.value
                node = node?.next
                return value
            }
            throw IllegalStateException()
        }

        override fun remove() {
        }
    }

    data class Node<T>(var value : T, var next : Node<T>?)
}
